protocol.go

```diff
@@ -371,43 +348,26 @@ func (p *Protocol) Close() {
 	}
 }
 
-func (p *Protocol) Run(transmit transmitFunc) {
-	ticker := time.NewTicker(p.updatePeriod)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-p.exit:
-			return
-		case <-ticker.C:
-			if err := p.retransmitUnackedPackets(transmit); err != nil && p.eh != nil {
-				p.eh(err)
-			}
-		}
+func (p *Protocol) checkIfRetransmit(idx uint16, resendTimeout time.Duration) ([]byte, bool) {
+	i := (p.oui + idx) % uint16(len(p.wq))
+	if p.wq[i] != uint32(p.oui+idx) || !p.wqe[i].shouldResend(time.Now(), resendTimeout) {
+		return nil, false
 	}
+	return p.wqe[i].buf.B, true
 }
 
-func (p *Protocol) retransmitUnackedPackets(transmit transmitFunc) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-
-	for idx := uint16(0); idx < uint16(len(p.wq)); idx++ {
-		i := (p.oui + idx) % uint16(len(p.wq))
-		if p.wq[i] != uint32(p.oui+idx) || !p.wqe[i].shouldResend(time.Now(), p.resendTimeout) {
-			continue
-		}
-
-		// log.Printf("%v: resend  (seq=%d)", &p, p.oui+idx)
-
-		if isEOF, err := transmit(p.wqe[i].buf.B); err != nil {
-			return fmt.Errorf("failed to retransmit unacked packet: %w", err)
-		} else if isEOF {
-			break
-		}
+func (p *Protocol) incrementWqe(idx uint16) {
+	i := (p.oui + idx) % uint16(len(p.wq))
+	p.wqe[i].written = time.Now()
+	p.wqe[i].resent++
+}
 
-		p.wqe[i].written = time.Now()
-		p.wqe[i].resent++
+func (p *Protocol) callErrorHandler(err error) {
```

Comment:
warning: exported method Protocol.Close should have comment or be unexported (golint)